BrainPortal/lib/globus_helpers.rb

```diff
@@ -240,25 +211,39 @@ def find_user_with_globus_identity(globus_identity)
   # account to the +identity+ provider. The array can
   # be empty (no such users) or contain more than one
   # user (an account management error).
-  def find_users_with_specific_identity(identity)
-    provider_id   = identity['identity_provider']              || cb_error("Globus: No identity provider")
-    provider_name = identity['identity_provider_display_name'] || cb_error("Globus: No identity provider name")
-    pref_username = identity['preferred_username'] ||
-                    identity['username']                       || cb_error("Globus: No preferred username")
+  def find_users_with_specific_identity(identity, oidc)
+    provider_id   = identity[oidc.identity_provider]              || cb_error("#{oidc.name}: No identity provider")
+    provider_name = identity[oidc.identity_provider_display_name] || cb_error("#{oidc.name}: No identity provider name")
+    pref_username = identity[oidc.preferred_username]             || cb_error("#{oidc.name}: No preferred username")
 
     # Special case for ORCID, because we already have fields for that provider
     if provider_name == 'ORCID'
       orcid = pref_username.sub(/@.*/, "")
       users = User.find_all_by_meta_data(:orcid, orcid).to_a
       return users if users.present?
-      # otherwise we fall through to detect users who linked with ORCID through Globus
+      # otherwise we fall through to detect users who linked with ORCID through OIDC
     end
 
     # All other globus providers
     # We need a user which match both the preferred username and provider_id
-    users = User.find_all_by_meta_data(:globus_preferred_username, pref_username)
+    users = User.find_all_by_meta_data(oidc.preferred_username_key, pref_username)
       .to_a
-      .select { |user| user.meta[:globus_provider_id] == provider_id }
+      .select { |user|  oidc.linked_provider_id(@user) == provider_id }
+  end
+
+  # Removes the recorded OIDC identity for +user+
+  def unlink_oidc_identity(oidc, user)
+    oidc.set_linked_provider_id(user, nil)
+    oidc.set_linked_provider_name(user, nil)
+    oidc.set_linked_preferred_username(user, nil)    
+    user.addlog("Unlinked #{oidc.name} identity")
+  end
+
+
+  def generate_oidc_login_uri(oidc_providers, redirect_url) #:nodoc:
+    @oidc_uris = {}
+    oidc_providers.each { |oidc| @oidc_uris[oidc.name] = oidc_login_uri(oidc, redirect_url)}
```

Comment:
When I went to `my account` page, this line (245) crashed because oidc_providers was `nil`

Reply:
The stack trace was:

```
[lib/globus_helpers.rb:245:in `generate_oidc_login_uri'](http://localhost:3000/users/2#)
[app/controllers/users_controller.rb:98:in `show'](http://localhost:3000/users/2#)
[app/controllers/application_controller.rb:115:in `activate_user_time_zone'](http://localhost:3000/users/2#)```

Reply:
It is because in the users controller code, method `show`, you misspelled one of the instance variables:

```ruby
    # Array of enabled OIDC providers configurations
    @oidc_configss = OidcConfig.enabled
    # Hash of OIDC uris with the OIDC name as key
    @oidc_uris      = generate_oidc_login_uri(@oidc_configs, globus_url)
```

(see the `@oidc_configss` instead of `@oidc_configs`... )